src/Validator/Rules/UniqueOperationTypes.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+declare(strict_types=1);
+
+namespace GraphQL\Validator\Rules;
+
+use GraphQL\Error\Error;
+use GraphQL\Language\AST\NodeKind;
+use GraphQL\Language\AST\SchemaDefinitionNode;
+use GraphQL\Language\AST\SchemaTypeExtensionNode;
+use GraphQL\Language\Visitor;
+use GraphQL\Language\VisitorOperation;
+use GraphQL\Validator\SDLValidationContext;
+
+/**
+ * Unique operation types
+ *
+ * A GraphQL document is only valid if it has only one type per operation.
+ */
+class UniqueOperationTypes extends ValidationRule
+{
+    public function getSDLVisitor(SDLValidationContext $context): array
+    {
+        $schema                 = $context->getSchema();
+        $definedOperationTypes  = [];
+        $existingOperationTypes = $schema !== null
+            ? [
+                'query' => $schema->getQueryType(),
+                'mutation' => $schema->getMutationType(),
+                'subscription' => $schema->getSubscriptionType(),
+            ]
+            : [];
+
+        /**
+         * @param SchemaDefinitionNode|SchemaTypeExtensionNode $node
+         */
+        $checkOperationTypes = static function ($node) use ($context, &$definedOperationTypes, $existingOperationTypes): VisitorOperation {
```

Comment:
```suggestion
        $checkOperationTypes = static function (Node $node) use ($context, &$definedOperationTypes, $existingOperationTypes): VisitorOperation {
```

Probably requires the hint above to use intersection types with `&Node`

Reply:
No `&Node` needed as `SchemaDefinitionNode` and `SchemaTypeExtensionNode` extends `Node`.

But the problem is, PHPStan doesn't understand `@params` on closures (https://github.com/phpstan/phpstan/issues/3770), therefore, without the `Node` type-hint, `$node` is `mixed`, with the type-hint, it's only `Node` resulting in `Access to an undefined property GraphQL\Language\AST\Node::$operationTypes.` error.

A possible solution is to check types explicitly:

```php
/**
 * @param SchemaDefinitionNode|SchemaTypeExtensionNode $node
 */
$checkOperationTypes = static function (Node $node) use ($context, &$definedOperationTypes, $existingOperationTypes): VisitorOperation {
    if (! $node instanceof SchemaDefinitionNode && ! $node instanceof SchemaTypeExtensionNode) {
        throw new TypeError('Expected $node to be ' . SchemaDefinitionNode::class . ' or ' . SchemaTypeExtensionNode::class . '.');
    }

    // ...
}
```

Once PHPStan starts understanding PhpDoc above closures, it can point this `if` as extraneous and we can remove it.

Reply:
You are right, `Node` is not needed. Thanks for researching that thoroughly.

I would rather forgo strict type safety if the alternative means worsening runtime performance. We have tests that ensure correctness, and its not like the code is going to run into a catastrophic issue like a memory leak if the types don't line up.

Let's just keep the code as is, that way we are ready for when PHPStan understands closure param hints. When we switch to PHP 8.1, we can just move to an actual type hint.